fontir/src/coords.rs

```diff
@@ -0,0 +1,315 @@
+//! Using confusable types for coords is an endless source of confusion; don't.
+//!
+//! See <https://github.com/googlefonts/fontmake-rs/blob/main/resources/text/units.md>
+
+use std::collections::BTreeMap;
+
+use log::warn;
+use ordered_float::OrderedFloat;
+use serde::{Deserialize, Serialize};
+
+use crate::piecewise_linear_map::PiecewiseLinearMap;
+
+/// A coordinate in some arbitrary space the designer dreamed up.
+///
+/// In .designspace, an xvalue. <https://fonttools.readthedocs.io/en/latest/designspaceLib/xml.html#dimension-element>.
+#[derive(Serialize, Deserialize, Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct DesignCoord(OrderedFloat<f32>);
+
+/// A coordinate the end user sees, e.g. what 'fvar' uses, Weight 400.
+///
+/// In .designspace, a uservalue. <https://fonttools.readthedocs.io/en/latest/designspaceLib/xml.html#dimension-element>.
+#[derive(Serialize, Deserialize, Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct UserCoord(OrderedFloat<f32>);
+
+/// A coordinate used within the font, not seen by any user.
+///
+/// Always in [-1, 1].
+///
+/// In .designspace, a uservalue. <https://fonttools.readthedocs.io/en/latest/designspaceLib/xml.html#dimension-element>.
+#[derive(Serialize, Deserialize, Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct InternalCoord(OrderedFloat<f32>);
+
+// Using BTreeMap instead of HashMap and OrderedFloat instead of f32 so that
+// the location is hashable and can be used as a key in Glyph::sources HashMap
+pub type DesignLocation = BTreeMap<String, DesignCoord>;
+pub type UserLocation = BTreeMap<String, UserCoord>;
+pub type InternalLocation = BTreeMap<String, UserCoord>;
+
+trait Converter<From, To> {
+    fn convert(&self, from: From) -> To;
+}
+
+/// Converts between Design, User, and Internal coordinates.
+///
+/// Stores [PiecewiseLinearMap]'s in several directions. Sources
+/// suggest <= 10 mappings is typical, we can afford the bytes.
+#[derive(Debug, Clone)]
+pub struct CoordConverter {
+    user_to_design: PiecewiseLinearMap,
+    design_to_user: PiecewiseLinearMap,
+    design_to_internal: PiecewiseLinearMap,
+    internal_to_design: PiecewiseLinearMap,
+}
+
+impl CoordConverter {
+    pub fn from_user_to_design_examples(
```

Comment:
I was confused to read "examples", I would call these (from, to) tuples just (axis) "mappings"... `from_user_to_design_axis_mappings`. It starts getting too verbose, how about just a default constructor `CoordConverter::new(mappings, default_idx)`, the type of the first parameter Vec<(UserCoord, DesignCoord)> already makes it explicit which direction it's expected to be, I don't think you will want to initialize it from anything else.

Reply:
Hm, agreed. I was stressing the potential for confusion but given that it's typed I'm +1 on `new(Vec<User,Design>)`.